optbuilder: take FOR SHARE locking during RC FK child checks

michae2 opened the pull request:
**logictest: fix fk_cascade_race_150282**

In #150291 we added subtest `fk_cascade_race_150282` which exercises FK
checks performed by concurrent Serializable and Read Committed
transactions. Unfortunately thanks to this test we've discovered that
the locking added in #150291 is insufficient for preventing all FK
violations. We need to turn on the disabled parent-FK-check locking for
the Serializable transaction.

This commmit fixes subtest `fk_cascade_race_150282` by turning on these
settings for the Serializable transaction:
- `enable_implicit_fk_locking_for_serializable`
- `enable_shared_locking_for_serializable`
- `enable_durable_locking_for_serializable`

Informs: #151663

Release note: None

---

**optbuilder: take FOR SHARE locking during RC FK child checks**

Under Read Committed isolation, we need to take FOR SHARE locking during
FK child checks to ensure that the rows read by the FK check have not
already been written to at a later timestamp. (This is the same reason
that we needed to add FOR UPDATE locking to RC FK cascades in #150291.)

Even with this locking, we still need Serializable transactions to have
the FK parent locking enabled, if both SSI and RC transactions are going
to concurrently modify rows involved in FK relationships. This is
because FOR SHARE (and FOR UPDATE) locking does not prevent phantoms.

Informs: #151663

Release note (bug fix): A bug that would allow FK violations as a result
of some combinations of concurrent Read Committed and Serializable
transactions has been fixed.

Note that if both Serializable and weaker-isolation transactions will
concurrently modify rows involved in FK relationships, the Serializable
transactions must have the following session variables set to prevent
any possible FK violations:

- `SET enable_implicit_fk_locking_for_serializable = on;`
- `SET enable_shared_locking_for_serializable = on;`
- `SET enable_durable_locking_for_serializable = on;`

michae2 committed:
logictest: fix fk_cascade_race_150282

In #150291 we added subtest `fk_cascade_race_150282` which exercises FK
checks performed by concurrent Serializable and Read Committed
transactions. Unfortunately thanks to this test we've discovered that
the locking added in #150291 is insufficient for preventing all FK
violations. We need to turn on the disabled parent-FK-check locking for
the Serializable transaction.

This commmit fixes subtest `fk_cascade_race_150282` by turning on these
settings for the Serializable transaction:
- `enable_implicit_fk_locking_for_serializable`
- `enable_shared_locking_for_serializable`
- `enable_durable_locking_for_serializable`

Informs: #151663

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152245)
<!-- Reviewable:end -->

michae2 committed:
optbuilder: take FOR SHARE locking during RC FK child checks

Under Read Committed isolation, we need to take FOR SHARE locking during
FK child checks to ensure that the rows read by the FK check have not
already been written to at a later timestamp. (This is the same reason
that we needed to add FOR UPDATE locking to RC FK cascades in #150291.)

Even with this locking, we still need Serializable transactions to have
the FK parent locking enabled, if both SSI and RC transactions are going
to concurrently modify rows involved in FK relationships. This is
because FOR SHARE (and FOR UPDATE) locking does not prevent phantoms.

Informs: #151663

Release note (bug fix): A bug that would allow FK violations as a result
of some combinations of concurrent Read Committed and Serializable
transactions has been fixed.

Note that if both Serializable and weaker-isolation transactions will
concurrently modify rows involved in FK relationships, the Serializable
transactions must have the following session variables set to prevent
any possible FK violations:

- `SET enable_implicit_fk_locking_for_serializable = on;`
- `SET enable_shared_locking_for_serializable = on;`
- `SET enable_durable_locking_for_serializable = on;`

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just a suggestion to improve a comment in the test

> @rytaft reviewed 1 of 1 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152245#-:-OYHMX1k-cbs5E4OFJXU:b48r5tz)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mw5h)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/fk_read_committed` line 314 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152245#-OYHKn1P-dg41g9B3dAb:-OYHKn1P-dg41g9B3dAc:bf3ubmj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b75ddfcac417210fac6291c8c0aca934f156c74a/pkg/ccl/logictestccl/testdata/logic_test/fk_read_committed#L314)):*
> ```
> #                                   commit
> #
> # Without proper locking we could end up with an orphan child row with p=1.
> ```

Can you say more here about what improper locking would look like and how that would lead to an orphan? Is this timeline showing the proper or the improper behavior?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152245)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mw5h, and @rytaft)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/fk_read_committed` line 314 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152245#-OYHKn1P-dg41g9B3dAb:-OYIrYB76rBtZ0xKfRPy:b-22fagq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b75ddfcac417210fac6291c8c0aca934f156c74a/pkg/ccl/logictestccl/testdata/logic_test/fk_read_committed#L314)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you say more here about what improper locking would look like and how that would lead to an orphan? Is this timeline showing the proper or the improper behavior?
</blockquote></details>

Yes, I fleshed out the comment a bit. This timeline shows the proper behavior.


<!-- Sent from Reviewable.io -->
